### PR TITLE
#69 - Improvements to Key class

### PR DESCRIPTION
**Pull request by olegmoz**
Closes issue #69
Implements all improvements proposed in the issue.

### CODE REVIEW

**Comment by olegmoz** [comment]
@g4s8 Please take a look at this PR and related issue #69

**Comment by paulodamaso** [comment]
@g4s8 Assign me please

**Comment by 0crat** [comment]
This pull request [#70](https://github.com/artipie/asto/issues/70) is assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/780483c9-8df6-4715-89af-76c8e3043a56); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; we should be aware that @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova) is on vacation; this ticket may be delayed; there will be a monetary reward for this job

File: src/test/java/com/artipie/asto/KeyTest.java [hunk 1]
```diff
@@ -48,4 +50,21 @@ void resovlesKeyFromBasePath() {
...
+            Matchers.equalTo(string)
+        );
+    }
+
+    @Test
+    void keyFromBadParts() {
+        Assertions.assertThrows(
+            Exception.class,
+            () -> new Key.From("part1", "part/2").string()
+        );
```
Comment (olenagerasimova) [inline src/test/java/com/artipie/asto/KeyTest.java]: @olegmoz I'd suggest to improve this assertion to check also exception class and message, smth like
```
MatcherAssert.assertThat(
            Assertions.assertThrows(
                IllegalStateException.class, 
                () -> new Key.From("part1", "part/2").string()
           ).getMessage(),
           new IsEqual<>("exception message")
)
```

File: src/main/java/com/artipie/asto/Key.java [hunk 2]
```diff
@@ -106,11 +113,16 @@ public From(final Key base, final String... parts) {
...
          */
         public From(final Collection<String> parts) {
-            this.parts = Collections.unmodifiableCollection(new ArrayList<>(parts));
+            this.parts = List.copyOf(parts);
         }
 
         @Override
         public String string() {
+            for (final String part : this.parts) {
+                if (part.contains("/")) {
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/asto/Key.java]: @olegmoz also I'd suggest to extract `/` into private constant, it's used 3 times in this class. Also it's odd that qulice is not complaining about duplicated literal...

File: src/test/java/com/artipie/asto/KeyTest.java (hunk 1 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/KeyTest.java]: @olenagerasimova sorry, but I do not see particular exception class or message to be part of method's contract right now, so I'd prefer to not enforce them in unit test

File: src/main/java/com/artipie/asto/Key.java (hunk 2 above)
Comment (olegmoz) [reply src/main/java/com/artipie/asto/Key.java]: @olenagerasimova thank you, extracted the constant. It seems that qualice ignores one symbol constants

**Comment by olegmoz** [comment]
@olenagerasimova made changes, please review

File: src/test/java/com/artipie/asto/KeyTest.java (hunk 1 above)
Comment (olenagerasimova) [reply src/test/java/com/artipie/asto/KeyTest.java]: @olegmoz hm, you are talking about `Key#string()` method contract, right? This method does not announce any exception in its contract, but you still wrote this test... Also, don't we write unit tests for particular implementations? The implementation of the `Key` this test is about throws `IllegalStateException` on line 128, why not to specifically test it?

Comment (olegmoz) [reply src/test/java/com/artipie/asto/KeyTest.java]: @olenagerasimova added checks to test

**Comment by olegmoz** [comment]
@olenagerasimova made changes by comment, please review

**Comment by olegmoz** [comment]
@g4s8 please review

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)[/z](https://www.0crat.com/u/olenagerasimova) didn't make CR comments

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**